app/worker/backend.go

```diff
@@ -0,0 +1,327 @@
+package worker
+
+import (
+	"crypto/tls"
+	"encoding/json"
+	"errors"
+	"io"
+	"net"
+	"net/url"
+	"os"
+	"os/exec"
+	"sync"
+	"time"
+
+	"github.com/gorilla/websocket"
+	"github.com/signal-golang/textsecure/rootCa"
+	log "github.com/sirupsen/logrus"
+)
+
+func RunRustBackend() {
+	var cmd *exec.Cmd
+	log.Infoln("[axolotl] Starting crayfish-backend")
+	if _, err := os.Stat("./crayfish"); err == nil {
+		cmd = exec.Command("./crayfish")
+	} else {
+		cmd = exec.Command("./backend/target/debug/crayfish")
+	}
+	var stdout, stderr []byte
+	var errStdout, errStderr error
+	stdoutIn, _ := cmd.StdoutPipe()
+	stderrIn, _ := cmd.StderrPipe()
+	err := cmd.Start()
+	if err != nil {
+		log.Fatalf("[axolotl] Starting crayfish-backend cmd.Start() failed with '%s'\n", err)
+	}
+	var wg sync.WaitGroup
+	wg.Add(1)
+	go func() {
+		stdout, errStdout = copyAndCapture(os.Stdout, stdoutIn)
+		wg.Done()
+	}()
+
+	stderr, errStderr = copyAndCapture(os.Stderr, stderrIn)
+
+	wg.Wait()
+
+	err = cmd.Wait()
+	if errStdout != nil || errStderr != nil {
+		log.Fatal("[axolotl] failed to capture stdout or stderr\n")
+	}
+	outStr, errStr := string(stdout), string(stderr)
+	log.Infof("\nout:\n%s\nerr:\n%s\n", outStr, errStr)
+	log.Infof("[axolotl] Crayfish-backend finished with error: %v", err)
+
+}
+func copyAndCapture(w io.Writer, r io.Reader) ([]byte, error) {
+	var out []byte
+	buf := make([]byte, 1024, 1024)
+	for {
+		n, err := r.Read(buf[:])
+		if n > 0 {
+			d := buf[:n]
+			out = append(out, d...)
+			_, err := w.Write(d)
+			if err != nil {
+				return out, err
+			}
+		}
+		if err != nil {
+			// Read returns io.EOF at the end of file, which is not an error for us
+			if err == io.EOF {
+				err = nil
+			}
+			return out, err
+		}
+	}
+}
+
+type CryfishWebSocketMessageType int32
+
+const (
+	CryfishWebSocketMessage_UNKNOWN  CryfishWebSocketMessageType = 0
+	CryfishWebSocketMessage_REQUEST  CryfishWebSocketMessageType = 1
+	CryfishWebSocketMessage_RESPONSE CryfishWebSocketMessageType = 2
+)
+
+type CryfishWebSocketMessage struct {
```

Comment:
Here we have the draft for the communication with cryfish:
```
{
  "type":1,
  "request":"",
  "response":""
}
```
Whereby 
```
	CryfishWebSocketMessage_UNKNOWN  CryfishWebSocketMessageType = 0
	CryfishWebSocketMessage_REQUEST  CryfishWebSocketMessageType = 1
	CryfishWebSocketMessage_RESPONSE CryfishWebSocketMessageType = 2
```

The request and the response fields structure could be like 
```
{
  "type":1,
  "message":"",
}
```
and the types defined like 
```
  CryfishWebSocketRequestMessageTyp_UNKNOWN              CryfishWebSocketRequestMessageType = 0
	CryfishWebSocketRequestMessageTyp_START_REGISTRATION   CryfishWebSocketRequestMessageType = 1
	CryfishWebSocketRequestMessageTyp_CONFIRM_REGISTRATION CryfishWebSocketRequestMessageType = 2
```
and the message field for 
For starting registration (aka 	CryfishWebSocketRequestMessageTyp_START_REGISTRATION == 1 )
```
{"number": "+491764315264", "password": "mypassword", "captcha": "123", "use_voice": false}
```

For confirming registration:

```
{"number": "+491764315264", "password": "mypassword", "confirm_code": 0, "signaling_key": [1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25,26,27,28,29,30,31,32,33,34,35,36,37,38,39,40,41,42,43,44,45,46,47,48,49,50,51,52]}

```

Reply:
Protocol should be fine for crayfish side. Only some error messages may still ignore the protocol...